crates/noirc_evaluator/src/lib.rs

```diff
@@ -191,22 +197,20 @@ impl Evaluator {
 
                 let mut struct_witnesses: BTreeMap<String, Vec<Witness>> = BTreeMap::new();
                 self.generate_struct_witnesses(&mut struct_witnesses, &new_fields)?;
-                if *visibility == AbiVisibility::Public {
-                    let witnesses: Vec<Witness> =
-                        struct_witnesses.values().flatten().cloned().collect();
-                    self.public_inputs.extend(witnesses);
-                }
-                ir_gen.abi_struct(name, Some(def), fields, struct_witnesses);
+
+                ir_gen.abi_struct(name, Some(def), fields, struct_witnesses.clone());
+                struct_witnesses.values().flatten().cloned().collect()
             }
             AbiType::String { length } => {
                 let typ = AbiType::Integer { sign: noirc_abi::Sign::Unsigned, width: 8 };
                 let witnesses = self.generate_array_witnesses(length, &typ)?;
-                if *visibility == AbiVisibility::Public {
-                    self.public_inputs.extend(witnesses.clone());
-                }
-                ir_gen.abi_array(name, Some(def), &typ, *length, witnesses);
+                ir_gen.abi_array(name, Some(def), &typ, *length, witnesses.clone());
+                witnesses
             }
-        }
+        };
+
+        self.param_witnesses.insert(name.to_owned(), witnesses);
```

Comment:
Makes sense -- so we no longer need to know whether a type from the main function is public or not, we simply parse it as a regular private type, then once we are done, we check the Abi for parameters which have been marked as public and fetch their witnesses to put in the public_inputs vector

Reply:
That's correct.